rrs/tools/rrs_upgrade_history.py

```diff
@@ -0,0 +1,293 @@
+#          Paul Eggleton <paul.eggleton@linux.intel.com>
+#
+# Licensed under the MIT license, see COPYING.MIT for details
+#
+# SPDX-License-Identifier: MIT
+
+import sys
+import os.path
+import optparse
+import logging
+from datetime import date, datetime, timedelta
+
+sys.path.insert(0, os.path.realpath(os.path.join(os.path.dirname(__file__))))
+from common import common_setup, get_logger
+
+common_setup()
+from layerindex import utils
+
+import git
+
+utils.setup_django()
+import settings
+
+logger = get_logger("HistoryUpgrade", settings)
+fetchdir = settings.LAYER_FETCH_DIR
+bitbakepath = os.path.join(fetchdir, 'bitbake')
+if not fetchdir:
+    logger.error("Please set LAYER_FETCH_DIR in settings.py")
+    sys.exit(1)
+
+
+def run_internal(maintplanlayerbranch, commit, commitdate, options, logger, bitbake_map, initial=False):
+    from layerindex.models import PythonEnvironment
+    from rrs.models import Release
+    if commitdate < maintplanlayerbranch.python3_switch_date:
+        # Python 2
+        if maintplanlayerbranch.python2_environment:
+            cmdprefix = maintplanlayerbranch.python2_environment.get_command()
+        else:
+            cmdprefix = 'python'
+        # Ensure we're using a bitbake version that is python 2 compatible
+        if commitdate > datetime(2016, 5, 10):
+            commitdate = datetime(2016, 5, 10)
+    else:
+        # Python 3
+        if maintplanlayerbranch.python3_environment:
+            cmdprefix = maintplanlayerbranch.python3_environment.get_command()
+        else:
+            cmdprefix = 'python3'
+
+    bitbake_rev = utils.runcmd(['git', 'rev-list', '-1', '--before=%s' % str(commitdate), 'origin/master'],
+                    bitbakepath, logger=logger)
+    check_rev = bitbake_map.get(bitbake_rev, None)
+    if check_rev:
+        logger.debug('Preferring bitbake revision %s over %s' % (check_rev, bitbake_rev))
+        bitbake_rev = check_rev
+
+    cmd = '%s upgrade_history_internal.py %s %s' % (cmdprefix, maintplanlayerbranch.layerbranch.id, commit)
+    if initial:
+        release = Release.get_by_date(maintplanlayerbranch.plan, commitdate)
+        if release:
+            comment = 'Initial import at %s release start.' % release.name
+        else:
+            comment = 'Initial import at %s' % commit
+        cmd += ' --initial="%s"' % comment
+    if bitbake_rev:
+        cmd += ' --bitbake-rev %s' % bitbake_rev
+    if options.filter_files:
+        cmd += ' --filter-files %s' % options.filter_files
+    if options.dry_run:
+        cmd += ' --dry-run'
+    if options.loglevel == logging.DEBUG:
+        cmd += ' --debug'
+    logger.debug('Running %s' % cmd)
+    ret, output = utils.run_command_interruptible(cmd)
+    if ret == 254:
+        # Interrupted by user, break out of loop
+        logger.info('Update interrupted, exiting')
+        sys.exit(254)
+
+"""
+    Upgrade history handler.
+"""
+def upgrade_history(options, logger):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 94 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 99% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this function or break it into multiple functions. For example, consider extracting the code block on lines 108-116 into a separate function.